### PR TITLE
smimesign 0.0.3 (new formula)

### PR DESCRIPTION
**Pull request by btoews**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

https://github.com/Homebrew/homebrew-core/pull/32003 was a previous attempt at submitting this formula. That PR was closed because the tool's repository was not notable enough. An adequate number of users have since starred the repository to meet the notability requirements.

/cc @fxcoudert @commitay who provided feedback on the original PR.

### CODE REVIEW

File: Formula/smimesign.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    (buildpath/"src/github.com/github/smimesign").install buildpath.children
+
+    cd "src/github.com/github/smimesign" do
+      system "go", "build", "-o", bin/"smimesign", "-ldflags", "-X main.versionString=#{version}"
+      prefix.install_metafiles
+    end
+  end
+
+  test do
+    assert_match version.to_s, shell_output("#{bin}/smimesign --version")
```
Comment (fxcoudert) [inline Formula/smimesign.rb]: We need a test that exercises the some of the functionality of the app.

File: Formula/smimesign.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  sha256 "69dfa16737c715679e067d0d929dd964790060b33981695d71bd6b047ae64e97"
+
+  depends_on "go" => :build
+
+  def install
+    ENV["GOPATH"] = buildpath
+    (buildpath/"src/github.com/github/smimesign").install buildpath.children
+
+    cd "src/github.com/github/smimesign" do
+      system "go", "build", "-o", bin/"smimesign", "-ldflags", "-X main.versionString=#{version}"
```
Comment (fxcoudert) [inline Formula/smimesign.rb]: Does this download dependencies? How are they specified? (I am always unsure, pinning dependencies in go-based formulas is always a bit tricky…)

**Review by fxcoudert — CHANGES_REQUESTED** [review]

File: Formula/smimesign.rb (hunk 1 above)
Comment (mastahyeti) [reply Formula/smimesign.rb]: The tool doesn't do much unless the user has a signing certificate and key in their keychain. We could just check the exit code from `smimesign --list-keys`, whose output may or may not be empty.

File: Formula/smimesign.rb (hunk 2 above)
Comment (mastahyeti) [reply Formula/smimesign.rb]: This tool vendors its dependencies, so nothing is downloaded at build time.

File: Formula/smimesign.rb (hunk 1 above)
Comment (fxcoudert) [reply Formula/smimesign.rb]: Yes, checking the output and exit code would be nice. Possibly try an operation with an invalid key and check that it fails as expected?

File: Formula/smimesign.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+      prefix.install_metafiles
+    end
+  end
+
+  test do
+    assert_match version.to_s, shell_output("#{bin}/smimesign --version")
+    system "#{bin}/smimesign", "--list-keys"
+    assert_match "could not find identity matching specified user-id: bad@identity", shell_output("#{bin}/smimesign -su bad@identity 2>&1", 1)
```
Comment (mastahyeti) [inline Formula/smimesign.rb]: I wasn't sure if there was a better helper for testing for a negative (non 0 exit) result.

**Comment by fxcoudert** [comment]
On El Capitan:

```
==> go build -o /usr/local/Cellar/smimesign/0.0.3/bin/smimesign -ldflags -X main.versionString=0.0.3
# github.com/github/smimesign/vendor/github.com/mastahyeti/certstore
vendor/github.com/mastahyeti/certstore/certstore_darwin.go:304:55: could not determine kind of name for C.SecKeyAlgorithm
vendor/github.com/mastahyeti/certstore/certstore_darwin.go:284:10: could not determine kind of name for C.SecKeyCreateSignature
vendor/github.com/mastahyeti/certstore/certstore_darwin.go:314:11: could not determine kind of name for C.kSecKeyAlgorithmECDSASignatureDigestX962SHA1
vendor/github.com/mastahyeti/certstore/certstore_darwin.go:316:11: could not determine kind of name for C.kSecKeyAlgorithmECDSASignatureDigestX962SHA256
vendor/github.com/mastahyeti/certstore/certstore_darwin.go:318:11: could not determine kind of name for C.kSecKeyAlgorithmECDSASignatureDigestX962SHA384
vendor/github.com/mastahyeti/certstore/certstore_darwin.go:320:11: could not determine kind of name for C.kSecKeyAlgorithmECDSASignatureDigestX962SHA512
vendor/github.com/mastahyeti/certstore/certstore_darwin.go:327:11: could not determine kind of name for C.kSecKeyAlgorithmRSASignatureDigestPKCS1v15SHA1
vendor/github.com/mastahyeti/certstore/certstore_darwin.go:329:11: could not determine kind of name for C.kSecKeyAlgorithmRSASignatureDigestPKCS1v15SHA256
vendor/github.com/mastahyeti/certstore/certstore_darwin.go:331:11: could not determine kind of name for C.kSecKeyAlgorithmRSASignatureDigestPKCS1v15SHA384
vendor/github.com/mastahyeti/certstore/certstore_darwin.go:333:11: could not determine kind of name for C.kSecKeyAlgorithmRSASignatureDigestPKCS1v15SHA512
```

**Comment by mastahyeti** [comment]
I'll look into whether its possible for smimesign to run on El Capitan. If not, is there some way to specify the oldest version of macOS that a formula works with? Maybe using the [`fails_with`](https://www.rubydoc.info/github/Homebrew/brew/master/Formula#fails_with-class_method) directive?

**Comment by fxcoudert** [comment]
`depends_on :macos => :sierra`

**Comment by mastahyeti** [comment]
It _does_ look like some of the APIs used by smimesign were added in 10.12. I won't be able to get things working on 10.11.

**Review by fxcoudert — APPROVED** [review]

**Comment by fxcoudert** [comment]
Thanks @mastahyeti for the nice pull request and the follow up. It is now merged!

**Comment by mastahyeti** [comment]
Thanks @fxcoudert!

**Pull request closed without merging**